text/0000-unsized-rvalues.md

```diff
@@ -0,0 +1,203 @@
+- Feature Name: unsized_locals
+- Start Date: 2017-02-11
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow for local variables, function arguments, and some expressions to have an unsized type, and implement it by storing the temporaries in variably-sized allocas.
+
+Have repeat expressions with a length that captures local variables be such an expression, returning an `[T]` slice.
+
+Provide some optimization guarantees that unnecessary temporaries will not create unnecessary allocas.
+
+# Motivation
+[motivation]: #motivation
+
+There are 2 motivations for this RFC:
+
+1. Passing unsized values, such as trait objects, to functions by value is often desired. Currently, this must be done through a `Box<T>` with an unnecessary allocation.
+
+  One particularly common example is passing closures that consume their environment without using monomorphization. One would like for this code to work:
+
+  ```Rust
+  fn takes_closure(f: FnOnce()) { f(); }
```

Comment:
It's going to be hard to explain to beginners why they should be writing `fn takes_closure<F: FnOnce()>(f: F)` vs. `fn takes_closure(f: FnOnce())`. The `fn takes_closure(f: FnOnce())` syntax is the easiest to read/write, IMO. However, I rarely want to pass unsized types by value or dynamically dispatch on owned types. It seems odd to grant the easiest syntax to the lesser-used feature.

I'm not sure of the best way to solve this. It might be a good idea to introduce `dyn`/`virtual` or a similar qualifier (cc @ticki) to prevent beginners from accidentally using dynamic dispatch everywhere. However, the fact that this doesn't mirror existing trait object syntax (e.g. `&FnOnce()` and `Box<FnOnce()>`) is regrettable.

P.S. Note that this problem would be made even worse with the addition of `fn takes_closure(f: impl FnOnce())` or `fn takes_closure(f: any FnOnce())`.